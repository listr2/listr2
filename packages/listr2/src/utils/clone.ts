import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
