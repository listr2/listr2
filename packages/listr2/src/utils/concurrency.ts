// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
