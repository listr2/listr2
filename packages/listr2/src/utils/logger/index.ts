export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
