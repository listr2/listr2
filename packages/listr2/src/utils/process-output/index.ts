export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
