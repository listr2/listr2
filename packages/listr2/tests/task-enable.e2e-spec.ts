import { Listr } from '@root'
import type { MockProcessOutput } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput } from '@tests/utils'

describe('enable with context', () => {
  const output: MockProcessOutput = {} as MockProcessOutput

  beforeEach(async() => {
    mockProcessOutput(output)
  })

  afterEach(async() => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  describe('should be disabled', () => {
    it('with a function returning boolean', async() => {
      await new Listr(
        [
          {
            title: 'disabled',
            enabled: (): boolean => false,
            task: (): Promise<void> => Promise.resolve()
          }
        ],
        { renderer: 'test' }
      ).run()

      expectProcessOutputToMatchSnapshot(output, 'aJWDY40NUyBBGenLqPWsXO6NOncHTUSU')
    })

    it('with async function returning boolean', async() => {
      await new Listr(
        [
          {
            title: 'disabled',
            enabled: async(): Promise<boolean> => Promise.resolve(false),
            task: (): Promise<void> => Promise.resolve()
          }
        ],
        { renderer: 'test' }
      ).run()

      expectProcessOutputToMatchSnapshot(output, 'QeoNcUfm5iAfZkWwttrc3l5dAB5LRvNx')
    })
  })

  describe('should be enabled', () => {
    it('with a function returning boolean', async() => {
      await new Listr(
        [
          {
            title: 'enabled',
            enabled: (): boolean => true,
            task: (): Promise<void> => Promise.resolve()
          }
        ],
        { renderer: 'test' }
      ).run()

      expectProcessOutputToMatchSnapshot(output, 'AebhVo1EK6MRCWEubovgNXHDo093g8YV')
    })

    it('with async function returning boolean', async() => {
      await new Listr(
        [
          {
            title: 'enabled',
            enabled: async(): Promise<boolean> => Promise.resolve(true),
            task: (): Promise<void> => Promise.resolve()
          }
        ],
        { renderer: 'test' }
      ).run()

      expectProcessOutputToMatchSnapshot(output, 've1QT0lP54yGe3x0e6yPYfb2SVEDulCE')
    })
  })

  describe('should be disabled depending on the context', () => {
    it('with an context variable', async() => {
      await new Listr(
        [
          {
            title: 'pre-task',
            task: (ctx): void => {
              ctx.enable = false
            }
          },
          {
            title: 'disabled',
            enabled: (ctx): boolean => ctx.enable,
            task: (): Promise<void> => Promise.resolve()
          }
        ],
        { renderer: 'test', concurrent: true }
      ).run()

      expectProcessOutputToMatchSnapshot(output, 'jbyappoYDm16wG6ARf5qOnS8CgSjhd7S')
    })

    it('with an injected context', async() => {
      const ctx = { enable: false }

      await new Listr(
        [
          {
            title: 'disabled',
            enabled: (ctx): boolean => ctx.enable,
            task: (): Promise<void> => Promise.resolve()
          }
        ],
        { renderer: 'test' }
      ).run(ctx)

      expectProcessOutputToMatchSnapshot(output, 'TkhbH3NQuuZN39gvRmwhH6iB6ZrNrxDX')
    })
  })
})
