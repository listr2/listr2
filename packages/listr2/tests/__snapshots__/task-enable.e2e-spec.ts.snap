// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`enable with context should be disabled depending on the context with an context variable: jbyappoYDm16wG6ARf5qOnS8CgSjhd7S-exit 1`] = `[]`;

exports[`enable with context should be disabled depending on the context with an context variable: jbyappoYDm16wG6ARf5qOnS8CgSjhd7S-stderr 1`] = `[]`;

exports[`enable with context should be disabled depending on the context with an context variable: jbyappoYDm16wG6ARf5qOnS8CgSjhd7S-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"pre-task","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["pre-task"]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"pre-task","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["pre-task"]}}
",
]
`;

exports[`enable with context should be disabled depending on the context with an injected context: TkhbH3NQuuZN39gvRmwhH6iB6ZrNrxDX-exit 1`] = `[]`;

exports[`enable with context should be disabled depending on the context with an injected context: TkhbH3NQuuZN39gvRmwhH6iB6ZrNrxDX-stderr 1`] = `[]`;

exports[`enable with context should be disabled depending on the context with an injected context: TkhbH3NQuuZN39gvRmwhH6iB6ZrNrxDX-stdout 1`] = `[]`;

exports[`enable with context should be disabled with a function returning boolean: aJWDY40NUyBBGenLqPWsXO6NOncHTUSU-exit 1`] = `[]`;

exports[`enable with context should be disabled with a function returning boolean: aJWDY40NUyBBGenLqPWsXO6NOncHTUSU-stderr 1`] = `[]`;

exports[`enable with context should be disabled with a function returning boolean: aJWDY40NUyBBGenLqPWsXO6NOncHTUSU-stdout 1`] = `[]`;

exports[`enable with context should be disabled with async function returning boolean: QeoNcUfm5iAfZkWwttrc3l5dAB5LRvNx-exit 1`] = `[]`;

exports[`enable with context should be disabled with async function returning boolean: QeoNcUfm5iAfZkWwttrc3l5dAB5LRvNx-stderr 1`] = `[]`;

exports[`enable with context should be disabled with async function returning boolean: QeoNcUfm5iAfZkWwttrc3l5dAB5LRvNx-stdout 1`] = `[]`;

exports[`enable with context should be enabled with a function returning boolean: AebhVo1EK6MRCWEubovgNXHDo093g8YV-exit 1`] = `[]`;

exports[`enable with context should be enabled with a function returning boolean: AebhVo1EK6MRCWEubovgNXHDo093g8YV-stderr 1`] = `[]`;

exports[`enable with context should be enabled with a function returning boolean: AebhVo1EK6MRCWEubovgNXHDo093g8YV-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"enabled","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["enabled"]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"enabled","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["enabled"]}}
",
]
`;

exports[`enable with context should be enabled with async function returning boolean: ve1QT0lP54yGe3x0e6yPYfb2SVEDulCE-exit 1`] = `[]`;

exports[`enable with context should be enabled with async function returning boolean: ve1QT0lP54yGe3x0e6yPYfb2SVEDulCE-stderr 1`] = `[]`;

exports[`enable with context should be enabled with async function returning boolean: ve1QT0lP54yGe3x0e6yPYfb2SVEDulCE-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"enabled","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["enabled"]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"enabled","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["enabled"]}}
",
]
`;
